cpp/tests/structs/structs_column_tests.cpp

```diff
@@ -623,4 +623,20 @@ TYPED_TEST(TypedStructColumnWrapperTest, CopyColumnFromView)
                                       cudf::column(list_of_structs_column->view()));
 }
 
+TEST_F(StructColumnWrapperTest, TestStructsColumnWithEmptyChild)
+{
+  // Structs should not superimpose their null mask onto any EMPTY children,
+  // because EMPTY columns cannot have a null mask. This test ensures that
+  // we can construct a structs column with a parent null mask and an EMPTY
+  // child.
+  auto empty_col =
+    std::make_unique<cudf::column>(cudf::data_type(cudf::type_id::EMPTY), 3, rmm::device_buffer{});
```

Comment:
This looks wrong to me. First, I would not call this an empty column. The type is empty (e.g. invalid). An an empty column normally has a valid type but no size. This one has invalid type with a size of 3. Seems like this should not be allowed -- we should throw exception for this condition. Why does the untyped column have a size? 

Is there a better way to test this? Perhaps the pytest is enough for this change.

Reply:
This is not  an "untyped" column, it's a column of type `EMPTY`. I always figured `EMPTY` to be the equivalent of Arrow's [`NullType`](https://arrow.apache.org/docs/cpp/api/datatype.html#classarrow_1_1_null_type), which is a way to represent "all null" columns with possibly non-zero length, using no memory at all.

Reply:
But even "all null" columns have types. And an empty column has a type too. 
What is the size of the child column in the pytest example?

Reply:
> But even "all null" columns have types

Yes, I agree: you could choose to represent an "all null" column with some other data type and set all its values to null. But Arrow chooses to use `NullType` when possible. And libcudf's Arrow reader maps `NullType` arrow arrays to  columns with`EMPTY`  data type.

In this example:

```
>>> s = pd.Series([[{"c": {"x": None}}], [{"c": None}]])
>>> s
0    [{'c': {'x': None}}]
1           [{'c': None}]
dtype: object
```

I consider this to be a list column with two elements. The values child of the list column is a structs column with two elements,  and _that_ structs column itself has one child.

Reply:
For arrow arrays the most inner level of the above array contains 2 elements of `NullType`.

Reply:
An _empty type_ column cannot have any nulls as per the exception check so an "all nulls" column with type `EMPTY` would not be possible, right? 
Would the test still work if the `empty_col` had 0 size?

Reply:
> the exception check
I'm slightly out of context here, which exception check?

Reply:
> an "all nulls" column with type EMPTY would not be possible, right?

Ah I see what you mean. By definition, it _is_ all nulls. Enabling it to have a null mask would mean introducing the ability to make some elements _non-null_, which is illegal. I believe this is why the [exception check](https://github.com/rapidsai/cudf/blob/branch-22.06/cpp/src/column/column_view.cpp#L51) exists.

---

> Would the test still work if the empty_col had 0 size?

Your question led to me correcting this test :-) What fails isn't _constructing_ the structs column, but attempting to `view()` it.

If the `empty_col` has 0 size, this test would pass on branch-22.06  as well, as there is no null mask in the parent that gets superimposed onto it.

Reply:
> What fails isn't constructing the structs column, but attempting to view() it.

To clarify, this is the exact cause with source links:

`make_structs_column` calls `superimpose_parent_nulls`.
https://github.com/rapidsai/cudf/blob/8d86ae88557053501535441518a17b4864cabcad/cpp/src/structs/structs_column_factories.cu#L45-L50

When a column view is created, this failure is triggered because the `EMPTY` column has been assigned a null mask pushed down from the parent structs column:
https://github.com/rapidsai/cudf/blob/8d86ae88557053501535441518a17b4864cabcad/cpp/src/column/column_view.cpp#L51